omniledger/service/contracts.go

```diff
@@ -138,6 +171,63 @@ func (s *Service) ContractConfig(cdb CollectionView, tx Instruction, coins []Coi
 // ContractDarc accepts the following instructions:
 //   - Spawn - creates a new darc
 //   - Invoke.Evolve - evolves an existing darc
-func (s *Service) ContractDarc(cdb CollectionView, tx Instruction, coins []Coin) (sc []StateChange, c []Coin, err error) {
-	return nil, nil, errors.New("Not yet implemented")
+func (s *Service) ContractDarc(coll CollectionView, tx Instruction,
+	coins []Coin) ([]StateChange, []Coin, error) {
+	if tx.Invoke == nil {
+		return nil, nil, errors.New("Darc can only be invoked (evolved)")
+	}
+	if tx.Invoke.Command == "_evolve" {
+		darcBuf := tx.Invoke.Args.Search("darc")
+		newD, err := darc.NewDarcFromProto(darcBuf)
+		if err != nil {
+			return nil, nil, err
+		}
+		newD.Signatures = append([]darc.Signature{}, tx.Signatures...)
+		// We create a a darc.GetDarc callback that looks into the
+		// collections database. If we are looking for the latest darc
+		// for some base ID, then LoadDarcFromColl will give it to us.
+		// Otherwise, we have to check whether the darc ID matches the
+		// query and only return if the darc if it does.
+		cb := func(s string, latest bool) *darc.Darc {
+			key, err := strToObjSlice(s)
+			if err != nil {
+				return nil
+			}
+			d, err := LoadDarcFromColl(coll, key)
+			if err != nil {
+				return nil
+			}
+			if latest {
+				return d
+			}
+			if s == darc.NewIdentityDarc(d.GetID()).String() {
+				return d
+			}
+			return nil
+		}
+		if err := newD.VerifyWithCB(cb, false); err != nil {
+			return nil, nil, err
+		}
+		return []StateChange{
+			NewStateChange(Update, tx.ObjectID, ContractDarcID, darcBuf),
+		}, nil, nil
+	} else if tx.Invoke.Command == "add" {
+		return nil, nil, errors.New("not implemented")
+	}
+	return nil, nil, errors.New("invalid command: " + tx.Invoke.Command)
+}
+
+func strToObjSlice(s string) ([]byte, error) {
+	ss := strings.Split(s, ":")
```

Comment:
maybe SplitN so that once the first colon is found it does not keep going? Cuts out a probably not real, but theoretical attack vector by giving a ginormous string with billions of colons in it.